### PR TITLE
Merge master into 1.x

### PR DESCRIPTION
**Pull request by nicolasmure**
Merge `master` into `1.x`.

Resolved conflicts and adapted the modifications of `master` to the usages of `1.x` (eg the `StorageFailure` exception usage)

## TODO
- [x] rebase the fixup commits

### CODE REVIEW

File: src/Gaufrette/Adapter/Local.php [hunk 1]
```diff
@@ -130,20 +138,34 @@ public function mtime($key)
...
             return;
         }
 
-        if (!unlink($this->computePath($key))) {
-            throw StorageFailure::unexpectedFailure('delete', ['key' => $key]);
-        }
+        throw StorageFailure::unexpectedFailure(
+            'delete',
+            ['key' => $key],
+            new FileNotFound($key)
```
Comment (NiR-) [inline src/Gaufrette/Adapter/Local.php]: You don't have to wrap `FileNotFound` exception in a `StorageFailure` actually. Both are valid exceptions accordingly to the Adapter interface (see [here](https://sourcegraph.com/github.com/KnpLabs/Gaufrette@1.x/-/blob/src/Gaufrette/Adapter.php#L84)). `StorageFailure` should be thrown only if others, more specific exceptions can't be thrown :wink:

File: src/Gaufrette/FilesystemMap.php [hunk 2]
```diff
@@ -23,76 +23,64 @@ public function all()
...
      */
-    public function get($domain)
+    public function get($name)
     {
-        if (!$this->has($domain)) {
+        if (!$this->has($name)) {
             throw new \InvalidArgumentException(sprintf(
-                'There is no filesystem defined for the "%s" domain.',
-                $domain
+                'There is no filesystem defined having "%s" name.',
```
Comment (NiR-) [inline src/Gaufrette/FilesystemMap.php]: `named` maybe ?

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,36 @@
+v0.8.3
```
[the remark this answers is not in the document]
Comment (nicolasmure) [reply CHANGELOG.md]: :+1: PRs merged in `1.x` should complete this indeed

File: spec/Gaufrette/Adapter/GridFSSpec.php [hunk 4]
```diff
@@ -162,4 +163,29 @@ function it_calculates_checksum($bucket)
...
     }
+
+    function it_throws_file_not_found_when_getting_the_size_of_unexisting_object($bucket)
+    {
+        $bucket->findOne(['filename' => 'do-no-exists'])->willThrow(FileNotFoundException::class);
+
+        $this->shouldThrow(FileNotFound::class)->duringSize('do-not-exists');
+    }
+
+    function it_should_throw_when_backend_do_not_specify_object_size($bucket)
```
Comment (NiR-) [inline spec/Gaufrette/Adapter/GridFSSpec.php]: `it_throws_...`

File: spec/Gaufrette/Adapter/GridFSSpec.php [hunk 5]
```diff
@@ -162,4 +163,29 @@ function it_calculates_checksum($bucket)
...
+    {
+        $bucket->findOne(['filename' => 'file'], Argument::any())
+            ->willReturn([])
+        ;
+
+        $this->shouldThrow(StorageFailure::class)->duringSize('file');
+    }
+
+    function it_should_retrieve_the_object_size($bucket)
```
Comment (NiR-) [inline spec/Gaufrette/Adapter/GridFSSpec.php]: `it_throws_...`

File: src/Gaufrette/Adapter/GridFS.php [hunk 6]
```diff
@@ -256,14 +256,21 @@ public function listKeys($prefix = '')
...
+     * {@inheritdoc}
+     */
     public function size($key)
     {
-        if (!$this->exists($key)) {
-            return false;
+        try {
+            $this->exists($key);
+        } catch (StorageFailure $e) {
+            throw new FileNotFound($key, $e->getCode(), $e);
```
Comment (NiR-) [inline src/Gaufrette/Adapter/GridFS.php]: `exists` can return a `StorageFailure`. In such case, it should not be wrapped by a `FileNotFound` exception. IMHO you should keep the condition and throw an exception in its body.

**Pull request merged**